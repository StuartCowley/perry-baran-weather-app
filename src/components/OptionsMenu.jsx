import React, { useState } from "react";
import PropTypes from "prop-types";
import "../styles/OptionsMenu.css";
import useUnitContext from "../hooks/useUnitContext";

function OptionsMenu({ handleSearch, setVisible }) {
  const { units, setUnits } = useUnitContext();
  const [selectedUnits, setSelectedUnits] = useState(units);
  const changeHandler = (e) => {
    setSelectedUnits(e.target.value);
  };

  const handleApply = (e) => {
    e.preventDefault();
    setUnits(selectedUnits);
    handleSearch(undefined, selectedUnits);
    setVisible(false);
  };

  return (
    <form className="options-menu-form" name="options menu form">
      <h2>Options</h2>
      <div className="options-menu-form__units">
        <h3>Units:</h3>
        <label htmlFor="standardRadio" className="options-menu-form__label">
          Scientific
          <input
            id="standardRadio"
            type="radio"
            value="standard"
            name="units"
            checked={selectedUnits === "standard"}
            onChange={changeHandler}
          />
        </label>
        <label htmlFor="metricRadio" className="options-menu-form__label">
          Metric
          <input
            id="metricRadio"
            type="radio"
            value="metric"
            name="units"
            checked={selectedUnits === "metric"}
            onChange={changeHandler}
          />
        </label>
        <label htmlFor="imperialRadio" className="options-menu-form__label">
          Imperial
          <input
            id="imperialRadio"
            type="radio"
            value="imperial"
            name="units"
            checked={selectedUnits === "imperial"}
            onChange={changeHandler}
          />
        </label>
      </div>
      <button
        type="submit"
        onClick={handleApply}
        className="options-menu-form__apply"
      >
        Apply
      </button>
    </form>
  );
}

OptionsMenu.propTypes = {
  handleSearch: PropTypes.func.isRequired,
  setVisible: PropTypes.func.isRequired,
};

export default OptionsMenu;
