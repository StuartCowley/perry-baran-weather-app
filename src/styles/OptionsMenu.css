.options-menu-form {
  display: flex;
  flex-direction: column;
  border-bottom: var(--top-bar-border);
  border-left: var(--top-bar-border);
  border-bottom-left-radius: 5px;  
  padding: 10px;
  background-color: var(--medium-blue);
  transition-duration: 500ms;
}

.options-menu-form h2 {
  font-size: 1.6rem;
  font-weight: 400;
  text-decoration: underline;
}

.options-menu-form__units {
  display: flex;
  align-items: center;
  margin: 10px 0;
}

.options-menu-form__units h3 {
  font-size: 1.2rem;
  margin-right: 5px;
}

.options-menu-form__label {
  width: 75px;
  padding: 5px 0;
  border: var(--border-dark-blue);
  border-left: hidden;
  color: var(--almost-white);
  text-align: center;
  cursor: pointer;
  transition-duration: 200ms;
}

.options-menu-form__label:first-of-type {
  border-left: var(--border-dark-blue);
}

.options-menu-form__label input {
  display: none;
}

.options-menu-form__label:has(input:checked) {
  background-color: var(--dark-blue);
}

.options-menu-form__label:hover {
  background-color: var(--dark-blue);
}

.options-menu-form__apply {
  all: unset;
  align-self: flex-end;
  padding: 5px 10px;
  border: var(--border-dark-blue);
  border-radius: 3px;
  color: white;
  cursor: pointer;
  transition-duration: 200ms;
}

.options-menu-form__apply:hover {
  background-color: var(--dark-blue);
}

