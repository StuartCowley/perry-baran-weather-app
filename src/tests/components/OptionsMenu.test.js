import React from "react";
import { render, fireEvent } from "@testing-library/react";
import OptionsMenu from "../../components/OptionsMenu";

const findRadio = (radios, value) => {
  return radios.find((radio) => radio.value === value);
};

describe("OptionsMenu", () => {
  const validProps = {
    handleSearch: jest.fn(),
    setVisible: jest.fn(),
  };
  const { handleSearch, setVisible } = validProps;
  let screen;

  beforeEach(() => {
    screen = render(
      <OptionsMenu handleSearch={handleSearch} setVisible={setVisible} />
    );
  });

  test("snapshot", () => {
    const { asFragment } = screen;

    expect(asFragment()).toMatchSnapshot();
  });

  test("all fields correctly rendered", () => {
    const { getByText, getAllByRole, getByRole } = screen;
    const radios = getAllByRole("radio");
    const standard = findRadio(radios, "standard");
    const metric = findRadio(radios, "metric");
    const imperial = findRadio(radios, "imperial");
    const apply = getByRole("button");

    expect(getByText(/options/i)).toBeInstanceOf(HTMLHeadingElement);
    expect(getByText(/units:/i)).toBeInstanceOf(HTMLHeadingElement);
    expect(radios).toHaveLength(3);
    expect(standard).toHaveAttribute("id", "standardRadio");
    expect(getByText(/scientific/i)).toHaveAttribute("for", "standardRadio");
    expect(metric).toHaveAttribute("id", "metricRadio");
    expect(getByText(/metric/i)).toHaveAttribute("for", "metricRadio");
    expect(imperial).toHaveAttribute("id", "imperialRadio");
    expect(getByText(/imperial/i)).toHaveAttribute("for", "imperialRadio");
    expect(apply).toHaveAttribute("type", "submit");
    expect(apply).toHaveTextContent(/apply/i);
  });

  test("selecting units", () => {
    const { getByRole, getAllByRole } = screen;
    const [standard, metric, imperial] = getAllByRole("radio");
    const apply = getByRole("button");

    fireEvent.click(standard);
    expect(standard).toBeChecked();
    expect(metric).not.toBeChecked();
    expect(imperial).not.toBeChecked();
    fireEvent.click(apply);
    expect(handleSearch).toBeCalledWith(undefined, standard.value);
    fireEvent.click(metric);
    expect(standard).not.toBeChecked();
    expect(metric).toBeChecked();
    expect(imperial).not.toBeChecked();
    fireEvent.click(apply);
    expect(handleSearch).toBeCalledWith(undefined, metric.value);
    fireEvent.click(imperial);
    expect(standard).not.toBeChecked();
    expect(metric).not.toBeChecked();
    expect(imperial).toBeChecked();
    fireEvent.click(apply);
    expect(handleSearch).toBeCalledWith(undefined, imperial.value);
  });

  test("apply button calls setVisible", () => {
    const { getByRole } = screen;
    const apply = getByRole("button");

    expect(setVisible).toBeCalledTimes(0);
    fireEvent.click(apply);
    expect(setVisible).toBeCalledWith(false);
  });
});
