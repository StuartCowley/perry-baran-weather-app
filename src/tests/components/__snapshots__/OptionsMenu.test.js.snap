// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OptionsMenu snapshot 1`] = `
<DocumentFragment>
  <form
    class="options-menu-form"
    name="options menu form"
  >
    <h2>
      Options
    </h2>
    <div
      class="options-menu-form__units"
    >
      <h3>
        Units:
      </h3>
      <label
        class="options-menu-form__label"
        for="standardRadio"
      >
        Scientific
        <input
          id="standardRadio"
          name="units"
          type="radio"
          value="standard"
        />
      </label>
      <label
        class="options-menu-form__label"
        for="metricRadio"
      >
        Metric
        <input
          checked=""
          id="metricRadio"
          name="units"
          type="radio"
          value="metric"
        />
      </label>
      <label
        class="options-menu-form__label"
        for="imperialRadio"
      >
        Imperial
        <input
          id="imperialRadio"
          name="units"
          type="radio"
          value="imperial"
        />
      </label>
    </div>
    <button
      class="options-menu-form__apply"
      type="submit"
    >
      Apply
    </button>
  </form>
</DocumentFragment>
`;
